kwok/ec2/ec2.go

```diff
@@ -0,0 +1,890 @@
+	"github.com/aws/smithy-go"
+	"github.com/docker/docker/pkg/namesgenerator"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/karpenter/kwok/apis/v1alpha1"
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/aws/karpenter-provider-aws/kwok/strategy"
+	"github.com/aws/karpenter-provider-aws/pkg/errors"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/instancetype"
+	"github.com/aws/karpenter-provider-aws/pkg/utils"
+)
+
+type Client struct {
+	ec2.Client
+	rateLimiterProvider RateLimiterProvider
+	kubeClient          client.Client
+	clock               clock.Clock
+
+	region        string
+	instanceTypes []ec2types.InstanceTypeInfo
+	subnets       []ec2types.Subnet
+	strategy      strategy.Strategy
+
+	instances sync.Map
+
+	launchTemplates        sync.Map
+	launchTemplateNameToID sync.Map
+}
+
+func NewClient(region string, ec2Client *ec2.Client, rateLimiterProvider RateLimiterProvider, strategy strategy.Strategy, kubeClient client.Client, clk clock.Clock, cfg *rest.Config) *Client {
+	var instanceTypes []ec2types.InstanceTypeInfo
+	instanceTypesPaginator := ec2.NewDescribeInstanceTypesPaginator(ec2Client, &ec2.DescribeInstanceTypesInput{
+		MaxResults: aws.Int32(100),
+	})
+	for instanceTypesPaginator.HasMorePages() {
+		output := lo.Must(instanceTypesPaginator.NextPage(context.Background()))
+		instanceTypes = append(instanceTypes, output.InstanceTypes...)
+	}
+	var subnets []ec2types.Subnet
+	subnetsPaginator := ec2.NewDescribeSubnetsPaginator(ec2Client, &ec2.DescribeSubnetsInput{
+		MaxResults: aws.Int32(100),
+	})
+	for subnetsPaginator.HasMorePages() {
+		output := lo.Must(subnetsPaginator.NextPage(context.Background()))
+		subnets = append(subnets, output.Subnets...)
+	}
+
+	c := &Client{
+		Client:              *ec2Client,
+		rateLimiterProvider: rateLimiterProvider,
+		kubeClient:          kubeClient,
+		clock:               clk,
+
+		region:        region,
+		instanceTypes: instanceTypes,
+		subnets:       subnets,
+		strategy:      strategy,
+
+		instances: sync.Map{},
+
+		launchTemplates:        sync.Map{},
+		launchTemplateNameToID: sync.Map{},
+	}
+	c.readBackup(context.Background(), cfg)
+	return c
+}
+
+func (c *Client) readBackup(ctx context.Context, cfg *rest.Config) {
+	configMaps := &corev1.ConfigMapList{}
+	lo.Must0(client.IgnoreNotFound(lo.Must(client.New(cfg, client.Options{})).List(ctx, configMaps, client.InNamespace("kube-system"))))
+
+	configMaps.Items = lo.Filter(configMaps.Items, func(c corev1.ConfigMap, _ int) bool {
+		return strings.Contains(c.Name, "kwok-aws-instances-")
+	})
```

Comment:
Is this faster than a bucket download on start? I guess definitely when reusing a cluster... Nevermind!

Reply:
It's also not just about speed -- I'm also trying to make it so that everything is localized to the cluster and ideally there are eventually no external infra requirements